pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,265 @@
+
+var supportedDistinguisherMethods = sets.NewString(
+	string(flowcontrol.FlowDistinguisherMethodByNamespaceType),
+	string(flowcontrol.FlowDistinguisherMethodByUserType),
+)
+
+var supportedVerbs = sets.NewString(
+	flowcontrol.VerbAll,
+	"get",
+	"list",
+	"watch",
+	"create",
+	"update",
+	"delete",
+	"deletecollection",
+	"patch",
+)
+
+var supportedSubjectKinds = sets.NewString(
+	flowcontrol.ServiceAccountKind,
+	flowcontrol.UserKind,
+	flowcontrol.GroupKind,
+)
+
+// ValidateFlowSchema validates the content of flow-schema
+func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&fs.ObjectMeta, false, ValidateFlowSchemaName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateFlowSchemaSpec(&fs.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateFlowSchemaStatus(&fs.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateFlowSchemaSpec validates the content of flow-schema's spec
+func ValidateFlowSchemaSpec(spec *flowcontrol.FlowSchemaSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.MatchingPrecedence < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("matchingPrecedence"), spec.MatchingPrecedence, "must be positive value"))
+	}
+	if spec.DistinguisherMethod != nil {
+		if !supportedDistinguisherMethods.Has(string(spec.DistinguisherMethod.Type)) {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("distinguisherMethod").Child("type"), spec.DistinguisherMethod, supportedDistinguisherMethods.List()))
+		}
+	}
+	if len(spec.PriorityLevelConfiguration.Name) > 0 {
+		for _, msg := range ValidatePriorityLevelConfigurationName(spec.PriorityLevelConfiguration.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("priorityLevelConfiguration").Child("name"), spec.PriorityLevelConfiguration.Name, msg))
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("priorityLevelConfiguration").Child("name"), "must reference to a priority level"))
+	}
+	if len(spec.Rules) > 0 {
+		for i, rule := range spec.Rules {
+			allErrs = append(allErrs, ValidateFlowSchemaPolicyRuleWithSubjects(&rule, fldPath.Child("rules").Index(i))...)
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("rules"), "rules must contain at least one value"))
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaPolicyRuleWithSubjects validates policy-rule-with-subjects object.
+func ValidateFlowSchemaPolicyRuleWithSubjects(rule *flowcontrol.PolicyRuleWithSubjects, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(rule.Subjects) > 0 {
+		for i, subject := range rule.Subjects {
+			allErrs = append(allErrs, ValidateFlowSchemaSubject(&subject, fldPath.Child("subjects").Index(i))...)
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("subjects"), "subjects must contain at least one value"))
```

Comment:
Doesn't match the doc on the API.  Seems like we'd want to allow empty.

Reply:
See my response there.

Reply:
i will remove `//+optional` to avoid confusion.